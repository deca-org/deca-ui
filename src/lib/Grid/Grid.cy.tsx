import { Test } from '../Utils';
import _cyp from '../../../cypress';
import React from 'react';
import Grid, { GridContainerProps, Cols } from '.';
import Box from '../Box';

const GridItemBox = ({ children }) => (
  <Box
    css={{
      bg: '$indigo500',
      color: '$white',
      width: '100%',
      p: '$1',
      textAlign: 'center',
      fontFamily: 'inherit',
    }}
  >
    {children}
  </Box>
);

const GridComposer = (props: GridContainerProps) => (
  <Grid.Container {...props} data-testid="test.gridContainer">
    <Grid data-testid="test.gridItem1">
      <GridItemBox>Item 1</GridItemBox>
    </Grid>
    <Grid data-testid="test.gridItem2">
      <GridItemBox>Item 2</GridItemBox>
    </Grid>
    <Grid data-testid="test.gridItem3">
      <GridItemBox>Item 3</GridItemBox>
    </Grid>
    <Grid data-testid="test.gridItem4">
      <GridItemBox>Item 4</GridItemBox>
    </Grid>
  </Grid.Container>
);

const OverridePreset = () => (
  <Grid.Container data-testid="test.gridContainer" xs={5}>
    <Grid data-testid="test.gridItem1" xs={6}>
      <GridItemBox>Item 1</GridItemBox>
    </Grid>
    <Grid data-testid="test.gridItem2">
      <GridItemBox>Item 2</GridItemBox>
    </Grid>
    <Grid data-testid="test.gridItem3" xs={7}>
      <GridItemBox>Item 3</GridItemBox>
    </Grid>
    <Grid data-testid="test.gridItem4" xs={2}>
      <GridItemBox>Item 4</GridItemBox>
    </Grid>
  </Grid.Container>
);

const ContainerBreakpointTest = () => (
  <Grid.Container
    data-testid="test.gridContainer"
    n={12}
    xs={6}
    sm={4}
    md={3}
    lg={2}
    xl={1}
  >
    <Grid data-testid="test.gridItem1">
      <GridItemBox>Item 1</GridItemBox>
    </Grid>
    <Grid data-testid="test.gridItem2">
      <GridItemBox>Item 2</GridItemBox>
    </Grid>
  </Grid.Container>
);

const BreakpointTest = () => (
  <Grid.Container data-testid="test.gridContainer">
    <Grid
      data-testid="test.gridItem1"
      n={12}
      xs={6}
      sm={4}
      md={3}
      lg={2}
      xl={1}
    >
      <GridItemBox>Item 1</GridItemBox>
    </Grid>
  </Grid.Container>
);

const containerSelector = '[data-testid="test.gridContainer"]';
const itemSelector = (itemNumber: number) =>
  `[data-testid="test.gridItem${itemNumber}"]`;

const parseFlexBasis = (flexBasisAmount: number) => {
  if (flexBasisAmount === 1) {
    return parseFloat(((flexBasisAmount / 12) * 100).toFixed(5)) + '%';
  } else {
    return parseFloat(((flexBasisAmount / 12) * 100).toFixed(4)) + '%';
  }
};

const containerColCheck = (breakpoint: string) => {
  // default amount of cols is 12 on each breakpoint
  const baseProps: Record<string, Cols> = {
    xs: 12,
    sm: 12,
    md: 12,
    lg: 12,
    xl: 12,
  };

  // iterate through column numbers 1-12
  for (let i = 1; i < 13; i++) {
    it(`${i} col(s) applied to all children`, () => {
      cy.viewport(Test.breakpoint(breakpoint), 500);
      // applying correct cols on target breakpoint
      const composerProps = { ...baseProps, [breakpoint]: i };
      cy.baseMount(<GridComposer {...composerProps} />);
      // iterate through how many grid items are in grid composer
      for (let j = 1; j < 5; j++) {
        cy.get(itemSelector(j)).should(
          'have.css',
          'flex-basis',
          parseFlexBasis(i)
        );
      }
    });
  }
};

const breakpointCSSCheck = () => {
  const allBreakpoints = ['n', 'xs', 'sm', 'md', 'lg', 'xl'];
  allBreakpoints.map((i) => {
    it(i, () => {
      const composerProps = { [i]: 6 };
      cy.viewport(Test.breakpoint(i), 500);
      cy.baseMount(
        <Grid.Container data-testid="test.gridContainer" {...composerProps}>
          <Grid
            data-testid="test.gridItem1"
            css={{
              [`@${i}`]: {
                fontFamily: 'Arial',
              },
            }}
          >
            <GridItemBox>Item 1</GridItemBox>
          </Grid>
        </Grid.Container>
      );
      cy.get(itemSelector(1)).should('have.css', 'font-family', 'Arial');
      cy.get(itemSelector(1)).should(
        'have.css',
        'flexBasis',
        parseFlexBasis(6)
      );
    });
  });
};

describe('components/Grid', () => {
  describe('flexbox properties', () => {
    it('flex-wrap', () => {
      cy.baseMount(<GridComposer />);
      cy.get(containerSelector).should('have.css', 'flex-wrap', 'wrap');
    });
    describe('justifyContent', () => {
      it('flex-start', () => {
        cy.baseMount(<GridComposer justifyContent="flex-start" />);
        cy.get(containerSelector).should(
          'have.css',
          'justifyContent',
          'flex-start'
        );
      });
      it('center', () => {
        cy.baseMount(<GridComposer justifyContent="center" />);
        cy.get(containerSelector).should(
          'have.css',
          'justifyContent',
          'center'
        );
      });
      it('flex-end', () => {
        cy.baseMount(<GridComposer justifyContent="flex-end" />);
        cy.get(containerSelector).should(
          'have.css',
          'justifyContent',
          'flex-end'
        );
      });
      it('space-between', () => {
        cy.baseMount(<GridComposer justifyContent="space-between" />);
        cy.get(containerSelector).should(
          'have.css',
          'justifyContent',
          'space-between'
        );
      });
      it('space-around', () => {
        cy.baseMount(<GridComposer justifyContent="space-around" />);
        cy.get(containerSelector).should(
          'have.css',
          'justifyContent',
          'space-around'
        );
      });
      it('space-evenly', () => {
        cy.baseMount(<GridComposer justifyContent="space-evenly" />);
        cy.get(containerSelector).should(
          'have.css',
          'justifyContent',
          'space-evenly'
        );
      });
    });
    describe('alignItems', () => {
      it('flex-start', () => {
        cy.baseMount(<GridComposer alignItems="flex-start" />);
        cy.get(containerSelector).should(
          'have.css',
          'alignItems',
          'flex-start'
        );
      });
      it('center', () => {
        cy.baseMount(<GridComposer alignItems="center" />);
        cy.get(containerSelector).should('have.css', 'alignItems', 'center');
      });
      it('flex-end', () => {
        cy.baseMount(<GridComposer alignItems="flex-end" />);
        cy.get(containerSelector).should('have.css', 'alignItems', 'flex-end');
      });
    });
  });
  describe('container col(s)', () => {
    describe('n', () => {
      containerColCheck('n');
    });
    describe('xs', () => {
      containerColCheck('xs');
    });
    describe('sm', () => {
      containerColCheck('sm');
    });
    describe('md', () => {
      containerColCheck('md');
    });
    describe('lg', () => {
      containerColCheck('lg');
    });
    describe('xl', () => {
      containerColCheck('xl');
    });
  });
  it('individual item props should override contianer props', () => {
    cy.viewport(Test.breakpoint('xs'), 500);
    cy.baseMount(<OverridePreset />);
    cy.get(itemSelector(1)).should('have.css', 'flex-basis', parseFlexBasis(6));
    cy.get(itemSelector(2)).should(
      'have.css',
      'flex-basis',
      parseFlexBasis(5) // use default container prop
    );
    cy.get(itemSelector(3)).should('have.css', 'flex-basis', parseFlexBasis(7));
    cy.get(itemSelector(4)).should('have.css', 'flex-basis', parseFlexBasis(2));
  });
  describe('container breakpoint testing', () => {
    it('n', () => {
      cy.viewport(Test.breakpoint('n'), 500);
      cy.baseMount(<ContainerBreakpointTest />);
      cy.get(itemSelector(1)).should(
        'have.css',
        'flex-basis',
        parseFlexBasis(12)
      );
      cy.get(itemSelector(2)).should(
        'have.css',
        'flex-basis',
        parseFlexBasis(12)
      );
    });
    it('xs', () => {
      cy.viewport(Test.breakpoint('xs'), 500);
      cy.baseMount(<ContainerBreakpointTest />);
      cy.get(itemSelector(1)).should(
        'have.css',
        'flex-basis',
        parseFlexBasis(6)
      );
      cy.get(itemSelector(2)).should(
        'have.css',
        'flex-basis',
        parseFlexBasis(6)
      );
    });
    it('sm', () => {
      cy.viewport(Test.breakpoint('sm'), 500);
      cy.baseMount(<ContainerBreakpointTest />);
      cy.get(itemSelector(1)).should(
        'have.css',
        'flex-basis',
        parseFlexBasis(4)
      );
      cy.get(itemSelector(2)).should(
        'have.css',
        'flex-basis',
        parseFlexBasis(4)
      );
    });
    it('md', () => {
      cy.viewport(Test.breakpoint('md'), 500);
      cy.baseMount(<ContainerBreakpointTest />);
      cy.get(itemSelector(1)).should(
        'have.css',
        'flex-basis',
        parseFlexBasis(3)
      );
      cy.get(itemSelector(2)).should(
        'have.css',
        'flex-basis',
        parseFlexBasis(3)
      );
    });
    it('lg', () => {
      cy.viewport(Test.breakpoint('lg'), 500);
      cy.baseMount(<ContainerBreakpointTest />);
      cy.get(itemSelector(1)).should(
        'have.css',
        'flex-basis',
        parseFlexBasis(2)
      );
      cy.get(itemSelector(2)).should(
        'have.css',
        'flex-basis',
        parseFlexBasis(2)
      );
    });
    it('xl', () => {
      cy.viewport(Test.breakpoint('xl'), 500);
      cy.baseMount(<ContainerBreakpointTest />);
      cy.get(itemSelector(1)).should(
        'have.css',
        'flex-basis',
        parseFlexBasis(1)
      );
      cy.get(itemSelector(2)).should(
        'have.css',
        'flex-basis',
        parseFlexBasis(1)
      );
    });
  });
  describe('individual breakpoint testing', () => {
    it('n', () => {
      cy.viewport(Test.breakpoint('n'), 500);
      cy.baseMount(<BreakpointTest />);
      cy.get(itemSelector(1)).should(
        'have.css',
        'flex-basis',
        parseFlexBasis(12)
      );
    });
    it('xs', () => {
      cy.viewport(Test.breakpoint('xs'), 500);
      cy.baseMount(<BreakpointTest />);
      cy.get(itemSelector(1)).should(
        'have.css',
        'flex-basis',
        parseFlexBasis(6)
      );
    });
    it('sm', () => {
      cy.viewport(Test.breakpoint('sm'), 500);
      cy.baseMount(<BreakpointTest />);
      cy.get(itemSelector(1)).should(
        'have.css',
        'flex-basis',
        parseFlexBasis(4)
      );
    });
    it('md', () => {
      cy.viewport(Test.breakpoint('md'), 500);
      cy.baseMount(<BreakpointTest />);
      cy.get(itemSelector(1)).should(
        'have.css',
        'flex-basis',
        parseFlexBasis(3)
      );
    });
    it('lg', () => {
      cy.viewport(Test.breakpoint('lg'), 500);
      cy.baseMount(<BreakpointTest />);
      cy.get(itemSelector(1)).should(
        'have.css',
        'flex-basis',
        parseFlexBasis(2)
      );
    });
    it('xl', () => {
      cy.viewport(Test.breakpoint('xl'), 500);
      cy.baseMount(<BreakpointTest />);
      cy.get(itemSelector(1)).should(
        'have.css',
        'flex-basis',
        parseFlexBasis(1)
      );
    });
  });
  describe('spacing', () => {
    it('none', () => {
      cy.baseMount(<GridComposer spacing="none" />);
      cy.get(containerSelector).should('have.css', 'margin', Test.space('n'));
      for (let i = 1; i < 5; i++) {
        cy.get(itemSelector(i)).should('have.css', 'padding', Test.space('n'));
      }
    });
    it('sm', () => {
      cy.baseMount(<GridComposer spacing="sm" />);
      cy.get(containerSelector).should('have.css', 'margin', Test.space('- 1'));
      for (let i = 1; i < 5; i++) {
        cy.get(itemSelector(i)).should('have.css', 'padding', Test.space('1'));
      }
    });
    it('md', () => {
      cy.baseMount(<GridComposer spacing="md" />);
      cy.get(containerSelector).should('have.css', 'margin', Test.space('- 3'));
      for (let i = 1; i < 5; i++) {
        cy.get(itemSelector(i)).should('have.css', 'padding', Test.space('3'));
      }
    });
    it('lg', () => {
      cy.baseMount(<GridComposer spacing="lg" />);
      cy.get(containerSelector).should('have.css', 'margin', Test.space('- 5'));
      for (let i = 1; i < 5; i++) {
        cy.get(itemSelector(i)).should('have.css', 'padding', Test.space('5'));
      }
    });
  });
  describe('custom css should not component breakpoints', () => {
    it('general css', () => {
      cy.viewport(Test.breakpoint('xs'), 500);
      cy.baseMount(
        <Grid.Container data-testid="test.gridContainer" n={12} xs={6} sm={3}>
          <Grid
            data-testid="test.gridItem1"
            css={{
              fontFamily: 'Arial',
            }}
          >
            <GridItemBox>Item 1</GridItemBox>
          </Grid>
        </Grid.Container>
      );
      cy.get(itemSelector(1)).should('have.css', 'font-family', 'Arial');
      cy.get(itemSelector(1)).should(
        'have.css',
        'flexBasis',
        parseFlexBasis(6)
      );
    });
    describe(
      'input breakpoints should not override system breakpoints',
      breakpointCSSCheck
    );
  });
});
