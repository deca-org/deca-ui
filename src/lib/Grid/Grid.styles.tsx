import { styled } from '@lib/Theme/stitches.config';
import { getStaticColor } from '@lib/Utils';
import { transparentize } from 'polished';

const justifyContentComposer = () => {
  const options = [
    'flex-start',
    'center',
    'flex-end',
    'space-between',
    'space-around',
    'space-evenly',
  ];
  const variantObj: Array<
    Array<keyof typeof options | Record<'justifyContent', keyof typeof options>>
  > = [];

  options.map((i) => {
    variantObj.push([
      i as keyof typeof options,
      { justifyContent: i } as Record<'justifyContent', keyof typeof options>,
    ]);
  });
  return Object.fromEntries(variantObj);
};

const alignItemsComposer = () => {
  const options = ['flex-start', 'center', 'flex-end'];
  const variantObj: Array<
    Array<keyof typeof options | Record<'alignItems', keyof typeof options>>
  > = [];

  options.map((i) => {
    variantObj.push([
      i as keyof typeof options,
      { alignItems: i } as Record<'alignItems', keyof typeof options>,
    ]);
  });
  return Object.fromEntries(variantObj);
};

export const StyledGridItem = styled('div', {
  display: 'block',
  boxSizing: 'border-box',
});

export const StyledGridContainer = styled('div', {
  display: 'flex',
  flexWrap: 'wrap',
  boxSizing: 'border-box',
  overflow: 'hidden',
  variants: {
    justifyContent: justifyContentComposer(),
    alignItems: alignItemsComposer(),
    spacing: {
      none: {
        m: '$n',
        [`& > ${StyledGridItem}`]: {
          p: '$n',
        },
      },
      sm: {
        m: '-$1',
        [`& > ${StyledGridItem}`]: {
          p: '$1',
        },
      },
      md: {
        m: '-$3',
        [`& > ${StyledGridItem}`]: {
          p: '$3',
        },
      },
      lg: {
        m: '-$5',
        [`& > ${StyledGridItem}`]: {
          p: '$5',
        },
      },
    },
  },
});

export const StyledGridRuler = styled('div', {
  position: 'absolute',
  top: 0,
  zIndex: -1,
  display: 'grid',
  gridTemplateColumns: 'repeat(12, 1fr)',
  width: '100%',
  height: '100vh',
  variants: {
    spacing: {
      none: {
        gap: '$n',
      },
      sm: {
        gap: 'calc($1 * 2)',
      },
      md: {
        gap: 'calc($3 * 2)',
      },
      lg: {
        gap: 'calc($5 * 2)',
      },
    },
  },
});

export const StyledGridRulerItem = styled('div', {
  bg: transparentize(0.75, getStaticColor('gray300')),
  border: `${transparentize(0.75, getStaticColor('gray600'))} solid 1px`,
});
