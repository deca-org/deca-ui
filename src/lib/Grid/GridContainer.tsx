import { CSS } from '@lib/Theme/stitches.config';
import { useDOMRef } from '@lib/Utils';
import clsx from 'clsx';
import React, { ReactElement } from 'react';

import { Cols, GridProps } from './Grid';
import { StyledGridContainer } from './Grid.styles';

/**
 * The GridContainer component serves as a wrapper component to the Grid component
 */
export interface GridContainerProps extends React.ComponentPropsWithRef<'div'> {
  /**
   * Changes which tag component outputs.
   */
  as?: keyof JSX.IntrinsicElements;
  /**
   * Override default CSS style.
   */
  css?: CSS;
  /**
   * The content of the component.
   */
  children?: React.ReactNode | undefined;
  /**
   * ClassName applied to the component.
   * @default ''
   */
  className?: string;
  /**
   * How much spacing there should be between columns.
   */
  spacing?: 'none' | 'sm' | 'md' | 'lg';
  /**
   * JustifyContent css prop.
   */
  justifyContent?:
    | 'flex-start'
    | 'center'
    | 'flex-end'
    | 'space-between'
    | 'space-around'
    | 'space-evenly';
  /**
   * AlignItems css prop.
   */
  alignItems?: 'flex-start' | 'center' | 'flex-end';
  /**
   * How many columns should be taken up by item initially
   */
  n?: Cols;
  /**
   * How many columns should be taken up by item on xs breakpoint
   */
  xs?: Cols;
  /**
   * How many columns should be taken up by item on sm breakpoint
   */
  sm?: Cols;
  /**
   * How many columns should be taken up by item on md breakpoint
   */
  md?: Cols;
  /**
   * How many columns should be taken up by item on lg breakpoint
   */
  lg?: Cols;
  /**
   * How many columns should be taken up by item on xl breakpoint
   */
  xl?: Cols;
}

const GridContainer = React.forwardRef(
  (
    {
      as = 'div',
      css,
      className = '',
      children,
      spacing = 'sm',
      justifyContent,
      alignItems,
      n,
      xs,
      sm,
      md,
      lg,
      xl,
      ...gridContainerProps
    }: GridContainerProps,
    ref: React.Ref<HTMLDivElement | null>
  ) => {
    const gridContainerRef = useDOMRef(ref);
    const preClass = 'decaGridContainer';

    return (
      <StyledGridContainer
        as={as}
        css={css}
        className={clsx(className, `${preClass}-root`)}
        ref={gridContainerRef}
        spacing={spacing}
        justifyContent={justifyContent}
        alignItems={alignItems}
        {...gridContainerProps}
      >
        {React.Children.map(
          children as React.ReactElement<GridProps>,
          (child: React.ReactElement<GridProps>) => {
            return React.cloneElement(child, {
              n: child.props.n ? child.props.n : n,
              xs: child.props.xs ? child.props.xs : xs,
              sm: child.props.sm ? child.props.sm : sm,
              md: child.props.md ? child.props.md : md,
              lg: child.props.lg ? child.props.lg : lg,
              xl: child.props.xl ? child.props.xl : xl,
            });
          }
        )}
      </StyledGridContainer>
    );
  }
);

export default GridContainer;
