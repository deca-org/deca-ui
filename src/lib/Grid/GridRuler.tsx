import React from 'react';
import { __DEV__ } from '@lib/Utils';
import { StyledGridRuler, StyledGridRulerItem } from './Grid.styles';

/**
 * The GridRuler component is meant to be used as a developer tool to ensure items are lined up correctly on a 12 column grid. For this reason, refs are not forwarded to this component.
 */
export interface Props {
  /**
   * How much spacing there should be between columns.
   */
  spacing?: 'none' | 'sm' | 'md' | 'lg';
}

const GridRuler = ({ spacing = 'sm' }: Props) => {
  return (
    <StyledGridRuler spacing={spacing}>
      {[1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12].map((i) => (
        <StyledGridRulerItem key={i} />
      ))}
    </StyledGridRuler>
  );
};

if (__DEV__) {
  GridRuler.displayName = 'DecaUI.GridRuler';
}

export default GridRuler;
