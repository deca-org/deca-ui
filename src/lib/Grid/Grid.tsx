import { CSS } from '@lib/Theme/stitches.config';
import {
  PolymorphicRef,
  PolymorphicComponentPropWithRef,
  __DEV__,
  MasterComponent,
} from '@lib/Utils';
import clsx from 'clsx';
import React from 'react';

import { StyledGridItem } from './Grid.styles';
import GridContainer from './GridContainer';

export type Cols = 1 | 2 | 3 | 4 | 5 | 6 | 7 | 8 | 9 | 10 | 11 | 12;

/**
 * The Grid component acts as a child to the GridContainer component
 */
interface Props {
  /**
   * Override default CSS style.
   */
  css?: CSS;
  /**
   * The content of the component.
   */
  children?: React.ReactNode | undefined;
  /**
   * ClassName applied to the component.
   * @default ''
   */
  className?: string;
  /**
   * How many columns should be taken up by item initially
   */
  n?: Cols;
  /**
   * How many columns should be taken up by item on xs breakpoint
   */
  xs?: Cols;
  /**
   * How many columns should be taken up by item on sm breakpoint
   */
  sm?: Cols;
  /**
   * How many columns should be taken up by item on md breakpoint
   */
  md?: Cols;
  /**
   * How many columns should be taken up by item on lg breakpoint
   */
  lg?: Cols;
  /**
   * How many columns should be taken up by item on xl breakpoint
   */
  xl?: Cols;
}

export type GridProps<T extends React.ElementType> =
  PolymorphicComponentPropWithRef<T, Props>;

export type GridComponent = (<C extends React.ElementType = 'div'>(
  props: GridProps<C>
) => React.ReactElement | null) & { displayName?: string };

const Grid: GridComponent = React.forwardRef(
  <T extends React.ElementType = 'div'>(
    {
      as,
      css,
      className = '',
      children,
      n,
      xs,
      sm,
      md,
      lg,
      xl,
      ...gridProps
    }: GridProps<T>,
    ref?: PolymorphicRef<T>
  ) => {
    const preClass = 'decaGrid';

    const genGridItemCss = (breakpoint?: Cols, bp?: CSS) => {
      if (bp) {
        return {
          flexBasis: `calc((${breakpoint} / 12) * 100%)`,
          maxWidth: `calc((${breakpoint} / 12) * 100%)`,
          ...bp,
        };
      }
      return {
        flexBasis: `calc((${breakpoint} / 12) * 100%)`,
        maxWidth: `calc((${breakpoint} / 12) * 100%)`,
      };
    };

    const {
      '@n': cssN,
      '@xs': cssXs,
      '@sm': cssSm,
      '@md': cssMd,
      '@lg': cssLg,
      '@xl': cssXl,
      ...otherCss
    } = (css as CSS) || {};

    const getCss = {
      flexGrow: 0,
      '@n': genGridItemCss(n, cssN),
      '@xs': genGridItemCss(xs, cssXs),
      '@sm': genGridItemCss(sm, cssSm),
      '@md': genGridItemCss(md, cssMd),
      '@lg': genGridItemCss(lg, cssLg),
      '@xl': genGridItemCss(xl, cssXl),
      ...otherCss,
    };

    return (
      <StyledGridItem
        as={as}
        css={getCss}
        className={clsx(className, `${preClass}-root`)}
        ref={ref}
        {...gridProps}
      >
        {children}
      </StyledGridItem>
    );
  }
);

if (__DEV__) {
  Grid.displayName = 'DecaUI.Grid';
}

export default Grid as MasterComponent<
  HTMLDivElement,
  GridProps<React.ElementType>,
  { Container: typeof GridContainer }
>;
