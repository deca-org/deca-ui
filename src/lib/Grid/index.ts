import Grid from './Grid';
import GridContainer from './GridContainer';

Grid.Container = GridContainer;

export * from './Grid';
export * from './GridContainer';
export { Grid };
export default Grid;
