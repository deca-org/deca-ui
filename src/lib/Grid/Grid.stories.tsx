import Box from '@lib/Box';
import Grid from '@lib/Grid';
import { ComponentMeta, ComponentStory } from '@storybook/react';
import React from 'react';

import GridRuler from './GridRuler';

export default {
  title: 'Grid',
  component: Grid,
  argTypes: {
    justifyContent: {
      description: 'JustifyContent css prop',
      table: {
        type: {
          summary:
            'flex-start | center | flex-end | space-between | space-around | space-evenly',
        },
      },
      control: {
        type: 'select',
        options: [
          'center',
          'flex-start',
          'flex-end',
          'space-between',
          'space-around',
          'space-evenly',
        ],
      },
    },
    alignItems: {
      description: 'AlignItems css prop',
      table: {
        type: {
          summary: 'flex-start | center | flex-end',
        },
      },
      control: {
        type: 'radio',
        options: ['flex-start', 'center', 'flex-end'],
      },
    },
    spacing: {
      description: 'How much spacing there should be between columns.',
      table: {
        type: {
          summary: 'none | sm | md | lg',
        },
      },
      control: {
        type: 'radio',
        options: ['none', 'sm', 'md', 'lg'],
      },
    },
  },
} as ComponentMeta<typeof Grid>;

const ExampleBox = ({ children }: { children: React.ReactNode }) => (
  <Box
    css={{
      bg: '$indigo600',
      p: '$3',
      color: '$white',
      textAlign: 'center',
      fontWeight: '$bold',
      border: '$indigo700 solid 1px',
      m: '$n',
      dflex: 'center',
    }}
  >
    {children}
  </Box>
);

const Template: ComponentStory<typeof Grid> = (args: any) => (
  <Box css={{ overflow: 'hidden' }}>
    <Grid.Container
      justifyContent={args.justifyContent}
      alignItems={args.alignItems}
      n={args.n}
      xs={args.xs}
      sm={args.sm}
      md={args.md}
      lg={args.lg}
      xl={args.xl}
      {...args}
    >
      <Grid>
        <ExampleBox>1</ExampleBox>
      </Grid>
      <Grid>
        <ExampleBox>2</ExampleBox>
      </Grid>
      <Grid>
        <ExampleBox>3</ExampleBox>
      </Grid>
      <Grid>
        <ExampleBox>4</ExampleBox>
      </Grid>
    </Grid.Container>
  </Box>
);

export const Default = Template.bind({});
(Default.args as any) = {
  justifyContent: 'center',
  alignItems: 'flex-start',
  spacing: 'sm',
  n: 12,
  xs: 12,
  sm: 6,
  md: 4,
  lg: 3,
  xl: 2,
};

export const WithGridRuler = Template.bind({});
WithGridRuler.args = { ...Default.args };

WithGridRuler.decorators = [
  (Story, context) => (
    <Box css={{ position: 'relative' }}>
      <Story />
      <GridRuler spacing={context.args.spacing} />
    </Box>
  ),
];
