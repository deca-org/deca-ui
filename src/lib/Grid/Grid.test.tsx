import Grid from '@lib/Grid';
import { render } from '@testing-library/react';
import '@testing-library/jest-dom';
import React from 'react';

describe('components/Grid', () => {
  it('renders content', () => {
    const { getByText } = render(
      <Grid.Container>
        <Grid>1</Grid>
        <Grid>2</Grid>
        <Grid>3</Grid>
      </Grid.Container>
    );
    expect(getByText('1')).toBeInTheDocument();
    expect(getByText('2')).toBeInTheDocument();
    expect(getByText('3')).toBeInTheDocument();
  });
  it('matches snapshot', () => {
    const { asFragment } = render(
      <Grid.Container>
        <Grid>1</Grid>
        <Grid>2</Grid>
        <Grid>3</Grid>
      </Grid.Container>
    );
    expect(asFragment()).toMatchSnapshot();
  });
  it('all sizes render properly on grid container', () => {
    const { asFragment } = render(
      <Grid.Container n={3} xs={4} sm={2} md={1} lg={12} xl={3}>
        <Grid>1</Grid>
        <Grid>2</Grid>
        <Grid>3</Grid>
      </Grid.Container>
    );
    expect(asFragment()).toMatchSnapshot();
  });
  it('all sizes render properly on grid item', () => {
    const { asFragment } = render(
      <Grid.Container>
        <Grid n={3} xs={4} sm={2} md={1} lg={12} xl={3}>
          1
        </Grid>
        <Grid>2</Grid>
        <Grid>3</Grid>
      </Grid.Container>
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
