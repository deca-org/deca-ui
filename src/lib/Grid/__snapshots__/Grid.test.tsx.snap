// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/Grid all sizes render properly on grid container 1`] = `
<DocumentFragment>
  <div
    class="c-kvxVqY c-kvxVqY-kUXYyw-spacing-sm decaGridContainer-root"
  >
    <div
      class="c-kuhmNX c-kuhmNX-igjvCRc-css decaGrid-root"
    >
      1
    </div>
    <div
      class="c-kuhmNX c-kuhmNX-igjvCRc-css decaGrid-root"
    >
      2
    </div>
    <div
      class="c-kuhmNX c-kuhmNX-igjvCRc-css decaGrid-root"
    >
      3
    </div>
  </div>
</DocumentFragment>
`;

exports[`components/Grid all sizes render properly on grid item 1`] = `
<DocumentFragment>
  <div
    class="c-kvxVqY c-kvxVqY-kUXYyw-spacing-sm decaGridContainer-root"
  >
    <div
      class="c-kuhmNX c-kuhmNX-igjvCRc-css decaGrid-root"
    >
      1
    </div>
    <div
      class="c-kuhmNX c-kuhmNX-ikaOtdY-css decaGrid-root"
    >
      2
    </div>
    <div
      class="c-kuhmNX c-kuhmNX-ikaOtdY-css decaGrid-root"
    >
      3
    </div>
  </div>
</DocumentFragment>
`;

exports[`components/Grid matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="c-kvxVqY c-kvxVqY-kUXYyw-spacing-sm decaGridContainer-root"
  >
    <div
      class="c-kuhmNX c-kuhmNX-ikaOtdY-css decaGrid-root"
    >
      1
    </div>
    <div
      class="c-kuhmNX c-kuhmNX-ikaOtdY-css decaGrid-root"
    >
      2
    </div>
    <div
      class="c-kuhmNX c-kuhmNX-ikaOtdY-css decaGrid-root"
    >
      3
    </div>
  </div>
</DocumentFragment>
`;
